design/automatic-network-configuration.md

```diff
@@ -0,0 +1,485 @@
+  networkConfigurationRef:
+  - name: nc1
+    namespace: default
+```
+
+#### BareMetalHost
+
+Add the `ports` field to `.spec` to indicate the connection information
+and performance of all network cards in the `BareMetalHost`.
+
+``` yaml
+spec:
+  ports:
+    - conPortID: 10
+      conDeviceRef:
+        name: switch0
+        kind: Switch
+        namespace: abc
+      capability:
+        speed: 1000
+        smartNic: false
+   - conPortID:
+   ......
+```
+
+#### CAPM3
+
+Add the following methods in `metal3machine_manager.go`:
+
+* `filter()`
+
+  Filter out the BMHs that cannot meet the network configuration requirements
+  from a list of BMHs (by calling the `Match()` function in the
+  NetworkConfiguration operator).
+  This `filter()` method will be used in the current `chooseHost()` method to
+  let `MachineManager` choose a `BMH` that meets the network requirements.
+
+* `configureNetwork()`
+
+  Create all the needed `NetworkBinding` CRs to trigger the network configuration
+  operation and check if the configuration is completed or not.
+
+* `deconfigureNetwork()`
+
+  Monitor the processing result after the `NetworkBinding` CR is deleted and return.
+
+### Add new CRD and controller
+
+#### NetworkConfiguration CRD
+
+// TODO: The content of the network configuration needs to be negotiated.
+
+// It is not clear what network configuration is required for other types
+
+// of network equipment.
+
+```yaml
+metaData:
+  name: nc1
+  finalizers:
+  - m3m1
+spec:
+  VLANs:
+    - id:
+    - id:
+  untaggedVLAN:
+  trunk:
+  # Indicates whether link aggregation is required. If the value is true,
+  # select two ports for link aggregation.
+  linkAggregation: false
+  linkAggregationType: # LAG, MLAG
+  # Network card performance required for this network configuration
+  nicHint:
+    speed: 1000
+    smartNIC: false
+  acl:
+    - type: # ipv4, ipv6
+      action: # allow, deny
+      protocol: # TCP, UDP, ICMP, ALL
+      src: # xxx.xxx.xxx.xxx/xx
+      srcPortRange: # 22, 22-30
+      des: # xxx.xxx.xxx.xxx/xx
+      desPortRange: # 22, 22-30
+status:
+  networkBindingRefs:
+    - name:
+      namespace:
+    - name:
+      namespace:
+```
+
+#### NetworkConfiguration Controller
+
+* User creates `NetworkConfiguration` CR
+  * The NetworkConfiguration Controller detects whether the data entered by the
+    user is legal, and adds `finalizers`.
+
+* User deletes `NetworkConfiguration` CR
+  * The NetworkConfiguration Controller detects whether the length of
+    `.status.networkBindingRefs` is 0, if it is not 0, it is not allowed to
+    delete the CR, otherwise the Controller deletes the `finalizers` and deletes
+    the CR.
+
+* User modify `NetworkConfiguration` CR
+  * The NetworkConfiguration Controller deletes all the old `NetworkBinding` CRs
+    and creates a new one for each deleted `NetworkBinding` CR according to the
+    contents in `.status.networkBindingRefs`.
+
+#### NetworkBinding CRD
```

Comment:
From what you put in the NetworkConfiguration and NetworkBinding, it looks like the two could be merged. What would be the value of keeping them split ?

Reply:
We do this to better divide permissions. Some users may only be able to create or modify NetworkBinding, but not modify NetworkConfiguration.

Reply:
We do this to better divide permissions. Some users may only be able to create or modify NetworkBinding, but not modify NetworkConfiguration.

Reply:
I don't think the user itself should create the network binding in most of the workflows. We should consider rather being able to reference the desired configuration maybe on the device for which it will be applied. that would remove the need for that intermediary object without changing the permissions requirements since they would not need modifications rights on the network configuration. CAPM3 could then set that reference when the workflow is automated.

Reply:
Our idea is to decouple this new operator from CAPM3 as much as possible, so that CAPM3 does not need to know the APIs of other types of devices.

Reply:
I agree with that, but CAPM3 will have to be aware of those types anyways, considering that it will be the one doing the mapping between the desired config at the cluster level, and the baremetal host (via the metal3machine(Template))

Reply:
If `CAPM3` is asked to modify the `configuration` reference used by each port, how to notify each device when the `configuration CR` is modified?

In our original design, we refer to the implementation principle of `deployment`. When the `configuration` is modified, all related `networkBinding` will be deleted and then created to trigger the device configuration update.

Reply:
Like I said above, I think the `NetworkBinding CRD`  should still be retained, so that when the `Configuration CR`  is modified, it can be very convenient to trigger the corresponding switch to modify the configuration.(By deleting and then creating  `NetworkBinding CR`). @maelk @zaneb @dhellmann WDYT?